### PR TITLE
FOUNDATIONS: Add Artist

### PR DESCRIPTION
**Pull request by NaynaMallick**

### CODE REVIEW

File: ArtGallery.Services.Api/Models/Artists/Exceptions/ArtistDependencyException.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// -----------------------------------------------------------------------
+
+using Xeptions;
+
+namespace ArtGallery.Services.Api.Models.Artists.Exceptions
+{
+    public class ArtistDependencyException : Xeption
+    {
+        public ArtistDependencyException(Xeption innerException)
+            : base(message: "Artist Dependency error occured, contract support.", innerException)
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Api/Models/Artists/Exceptions/ArtistDependencyException.cs]: small d

File: ArtGallery.Services.Api/Models/Artists/Exceptions/ArtistValidationException.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+// -----------------------------------------------------------------------
+
+using Xeptions;
+
+namespace ArtGallery.Services.Api.Models.Artists.Exceptions
+{
+    public class ArtistValidationException : Xeption
+    {
+        public ArtistValidationException(Xeption innerException)
+            : base(message: "Please fix the error and try again.", innerException)
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Api/Models/Artists/Exceptions/ArtistValidationException.cs]: Messaging to be more specific - Artist validation error occurred.

File: ArtGallery.Services.Api/Models/Artists/Exceptions/FailedArtistStorageException.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+using System;
+using Xeptions;
+
+namespace ArtGallery.Services.Api.Models.Artists.Exceptions
+{
+    public class FailedArtistStorageException : Xeption
+    {
+        public FailedArtistStorageException(Exception innerException)
+            : base(message: "Failed Artist Storage error occured.", innerException)
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Api/Models/Artists/Exceptions/FailedArtistStorageException.cs]: small letter - artist and storage

File: ArtGallery.Services.Api/Services/Foundations/Artists/ArtistService.Exceptions.cs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        {
+            var artistDependencyException = new ArtistDependencyException(exception);
+            this.loggingBroker.LogCritical(artistDependencyException);
+
+            return artistDependencyException;
+        }
+
+        private ArtistDependencyValidationException CreateAndLogDependencyValidationException(Xeption exception)
+        {
+            var artistValidationDependencyException = new ArtistDependencyValidationException(exception);
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Api/Services/Foundations/Artists/ArtistService.Exceptions.cs]: Correct the variable name

File: ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+            //then
+            await Assert.ThrowsAsync<ArtistDependencyValidationException>(() =>
+                addArtistTask.AsTask());
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTime(),
+                    Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(expectedArtistDepdendencyValidationException))),
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs]: Move the variable to next line.

File: ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+        [Fact]
+        public async void ShouldThrowDependencyExceptionOnAddIfDatabaseUpdateErrorOccursAndLogItAsync()
+        {
+            // given
+            Artist someArtist = CreateRandomArtist();
+            string randomMessage = GetRandomString();
+            var databaseUpdateException = new DbUpdateException();
+
+            var failedStorageDependencyException =
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs]: Update variable name

File: ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+            var expectedArtistDependencyException =
+                new ArtistDependencyException(failedArtistStorageException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTime())
+                    .Throws(sqlException);
+
+            //when
+            ValueTask<Artist> addArtistTask =
+                this.artistService.AddArtistAsync(randomArtist);
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs]: someArtist is preferred to keep it consistent

File: ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Validations.Add.cs [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+            ValueTask<Artist> addArtistTask =
+                this.artistService.AddArtistAsync(nullArtist);
+
+            //then
+            await Assert.ThrowsAsync<ArtistValidationException>(() =>
+                addArtistTask.AsTask());
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(
+                    SameExceptionAs(expectedArtistValidationException))),
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Validations.Add.cs]: Move 'SameExceptionAs' to above

File: ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Validations.Add.cs [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+                        Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.InsertArtistAsync(It.IsAny<Artist>()),
+                    Times.Never);
+
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Validations.Add.cs:44]: Missing date time broker. Check in other tests as well.

File: ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        private static Artist CreateRandomArtist() =>
+            CreateArtistFiller(dateTime: GetRandomDateTime()).Create();
+
+        private static Artist CreateRandomArtist(DateTimeOffset dateTime) =>
+            CreateArtistFiller(dateTime).Create();
+
+        private static SqlException GetSqlException() =>
+            (SqlException)FormatterServices.GetUninitializedObject(typeof(SqlException));
+
+        private static Expression<Func<Xeption, bool>> SameExceptionAs(Exception expectedException)
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.cs]: Parameter type - Xeption

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

File: ArtGallery.Services.Api/Models/Artists/Exceptions/ArtistDependencyException.cs (hunk 1 above)
Comment (NaynaMallick) [reply ArtGallery.Services.Api/Models/Artists/Exceptions/ArtistDependencyException.cs]: fixed

File: ArtGallery.Services.Api/Models/Artists/Exceptions/ArtistValidationException.cs (hunk 2 above)
Comment (NaynaMallick) [reply ArtGallery.Services.Api/Models/Artists/Exceptions/ArtistValidationException.cs]: fixed

File: ArtGallery.Services.Api/Models/Artists/Exceptions/FailedArtistStorageException.cs (hunk 3 above)
Comment (NaynaMallick) [reply ArtGallery.Services.Api/Models/Artists/Exceptions/FailedArtistStorageException.cs]: fixed

File: ArtGallery.Services.Api/Services/Foundations/Artists/ArtistService.Exceptions.cs (hunk 4 above)
Comment (NaynaMallick) [reply ArtGallery.Services.Api/Services/Foundations/Artists/ArtistService.Exceptions.cs]: done

File: ArtGallery.Services.Api/Startup.cs [hunk 11]
```diff
@@ -64,9 +66,14 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
         {
+            services.AddTransient<IStorageBroker, StorageBroker>();
             services.AddTransient<IDateTimeBroker, DateTimeBroker>();
             services.AddTransient<ILoggingBroker, LoggingBroker>();
-            services.AddTransient<IStorageBroker, StorageBroker>();
+        }
+
+        private static void AddServices(IServiceCollection services)
+        {
+            services.AddTransient<IArtistService, ArtistService>();
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Api/Startup.cs:75]: Single lie so you can do it =>

File: ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+                addArtistTask.AsTask());
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTime(),
+                    Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                   expectedArtistDepdendencyException))),
+                    Times.Once);
```
Comment (ShriHumrudha) [inline ArtGallery.Services.Tests.Unit/Services/Foundations/Artists/ArtistServiceTests.Exceptions.Add.cs]: Tab needed here

**Review by ShriHumrudha — CHANGES_REQUESTED** [review]

**Review by ShriHumrudha — APPROVED** [review]

**Pull request merged**